sympy/codegen/ast.py

```diff
@@ -399,3 +400,405 @@ def body(self):
         Must be an iterable object or CodeBlock.
         """
         return self._args[2]
+
+
+class Type(Symbol):
+    """ Represents a type.
+
+    The naming is a super-set of NumPy naming, see [1]_.
+
+    Arguments
+    ---------
+    name : str or Type
+        Either an explicit type: ``intc``, ``intp``, ``int8``, ``int16``,
+        ``int32``, ``int64``, ``uint8``, ``uint16``, ``uint32``, ``uint64,
+        float16``, ``float32``, ``float64``, ``complex64``, ``complex128``,
+        ``bool. Or only kind (precision decided by code-printer): ``integer``,
+        ``real`` or ``complex`` (where the latter two are of floating point type).
+        If a ``Type`` instance is given, the said instance is returned.
+
+    Examples
+    --------
+    >>> from sympy.codegen.ast import Type
+    >>> Type.from_expr(42).name
+    'integer'
+    >>> f32 = Type("float32")
+    >>> v6 = 0.123456
+    >>> f32.cast_check(v6)
+    0.123456
+    >>> v10 = 12345.67894
+    >>> f32.cast_check(v10)  # doctest: +ELLIPSIS
+    Traceback (most recent call last):
+      ...
+    ValueError: Casting gives a significantly different value.
+    >>> f64 = Type('float64')
+    >>> f64.cast_check(v10)
+    12345.67894
+    >>> from sympy import Float
+    >>> v18 = Float('0.123456789012345646')
+    >>> f64.cast_check(v18)
+    Traceback (most recent call last):
+      ...
+    ValueError: Casting gives a significantly different value.
+    >>> Type('float80').cast_check(v18)
+    0.123456789012345649
+
+    References
+    ----------
+    [1] https://docs.scipy.org/doc/numpy/user/basics.types.html
+
+    """
+    allowed_names = tuple('intc intp int8 int16 int32 int64 uint8 uint16 uint32'.split() +
```

Comment:
Maybe instead of checking in the constructor, we can just have a set of known types as constants in this file, like

```
intc = Type('intc')
intp = Type('intp')
...
```